File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 142085. 디펜스 게임/kotlin/Solution.kt
class Solution {
    fun solution(n: Int, k: Int, enemy: IntArray): Int {
        var life = n
        var chance = k
        val heap = Heap(enemy.size)
        for (i in enemy.indices) {
            life -= enemy[i]
            heap.add(enemy[i])
            if (life < 0) {
                if (chance > 0 && heap.isNotEmpty()) {
                    life += heap.poll()
                    chance --
                } else {
                    return i
                }
            }
        }
        return enemy.size
    }
}

class Heap(cap: Int) {

    private val arr = IntArray(cap + 1)

    private var size = 0

    fun add(e: Int) {
        arr[++size] = e
        var i = size
        var p = size shr 1
        var big = i
        while (p > 0) {
            if (arr[i] > arr[p]) {
                arr[p] = arr[i].also { arr[i] = arr[p] }
                i = p
                p = i shr 1
            } else {
                break
            }
        }
    }

    fun poll() : Int {
        val e = arr[1]
        arr[1] = arr[size--]
        var i = 1
        var lt = i shl 1
        var rt = lt + 1
        var big = i
        while(true) {
            if (lt <= size && arr[lt] > arr[big]) big = lt
            if (rt <= size && arr[rt] > arr[big]) big = rt
            if (i == big) break
            arr[i] = arr[big].also{arr[big] = arr[i]}
            i = big
            lt = i shl 1
            rt = lt + 1
        }
        return e
    }

    fun isNotEmpty() = size != 0

}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 142085. 디펜스 게임/python/solution.py
from heapq import heappush, heappop


def solution(n, k, enemy):
    heap = []

    for i, e in enumerate(enemy):
        # 현재 적의 수만큼 차감하고, 최대힙에 삽입
        n -= e
        heappush(heap, -e)

        # 적의 수가 더 많아서 n 이 0보다 작아졌을 경우
        if n < 0:
            if k > 0:
                # 무적권이 있다면, 무적권을 사용하고 heap 에서 최댓값을 찾아 그만큼 n을 복원
                n += - heappop(heap)
                k -= 1
            else:
                # 무적권이 없다면, 라운드 값을 반환
                return i
    return len(enemy)  # 끝까지 살았으면 enemy 배열 길이를 반환

